From ed329e0e9ad0819ad32e18ec3f4b283caf34a5ee Mon Sep 17 00:00:00 2001
From: Siva Varalakshmi Devi <sivavaralakshmidevi@ami.com>
Date: Tue, 3 Dec 2024 15:14:03 +0530
Subject: [PATCH 14/14] OT-8821 Handle Power-Saving Mode via Virtual Media

This patch allows to enable and disable the power-saving mode via VM service
when media redirection starts and stops before registering the session and
after unregistering the session information.

Signed-off-by: Siva Varalakshmi Devi <sivavaralakshmidevi@ami.com>
---
 src/system.hpp | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/src/system.hpp b/src/system.hpp
index 49a9798..8d7230f 100644
--- a/src/system.hpp
+++ b/src/system.hpp
@@ -20,6 +20,9 @@
 
 namespace fs = std::filesystem;
 
+#define POWER_SAVE_MODE_ENABLE 1
+#define POWER_SAVE_MODE_DISABLE 0
+
 /* Map to keep track of Active session and bind [Slot_N - sessionID] */
 using activesessionmap = std::map<std::string, uint8_t>;
 static activesessionmap activeSessons;
@@ -694,6 +697,35 @@ class DbusMonitor
     }
 };
 
+inline void powerSaveMode(int status)
+{
+    if ((status == 0) || (status == 1))
+    {
+        try
+        {
+            auto bus = sdbusplus::bus::new_system();
+            auto methodCall = bus.new_method_call(
+                "xyz.openbmc_project.Settings",
+                "/xyz/openbmc_project/logging/settings",
+                "xyz.openbmc_project.USB", "SetUSBPowerSaveMode");
+            methodCall.append(status);
+            bus.call(methodCall);
+        }
+
+        catch (const sdbusplus::exception::SdBusError& e)
+        {
+            LogMsg(Logger::Error,"D-Bus call Failed ERROR=%s", e.what());
+            return;
+        }
+
+        catch (const std::exception& e)
+        {
+            LogMsg(Logger::Error,"Error handling for powersavemode=%s", e.what());
+            return;
+        }
+    }
+}
+
 /* Returns slot number if ejected, -1 otherwise */
 static int eject_status(const std::string& filePath)
 {
@@ -798,6 +830,7 @@ struct UsbGadget
         {
             try
             {
+                powerSaveMode(POWER_SAVE_MODE_DISABLE);
                 fs::create_directories(gadgetDir);
                 echoToFile(gadgetDir / "idVendor", "0x1d6b");
                 echoToFile(gadgetDir / "idProduct", "0x0104");
@@ -1082,6 +1115,7 @@ struct UsbGadget
                 LogMsg(Logger::Error, "failed to Unregister Session");
             }
             activeSessons.erase(name);
+            powerSaveMode(POWER_SAVE_MODE_ENABLE);
         }
 
         const std::array<const char*, 6> dirs = {
-- 
2.34.1

